BLD: start building wheels against numpy 2.0 [wheel build]

andyfaff opened the pull request:
This PR adjusts the wheel build process to build against numpy2.0. This is being done so that downstream users can start ferreting out possible issues in the lead-up to numpy 2.0.

@seberg 
@rgommers 

Also @tylerjereddy for visibility as release manager.

andyfaff commented:
Ok, I started building this on my fork, only running the wheel building. The wheel building works after bumping the numpy versions in pyproject.toml, but the rest of CI will be broken. I need to think a bit more.

rgommers commented:
> but the rest of CI will be broken.

Can you explain why that would be the case? I'd expect it to work without problems for the full range of numpy versions we support as a runtime dependency.

andyfaff commented:
In [78dfe67](https://github.com/scipy/scipy/pull/19217/commits/78dfe67a2c84a5b40791d134630dae3206f8def3) I adjusted the `build-system.requires` in pyproject.toml to require numpy>=2.0.0.dev0. This was fine for wheel build because made adjustments to get numpy2 from the wheels bucket. However, it made other CI runs fail (e.g. https://github.com/scipy/scipy/actions/runs/6128741954/job/16636113234), because they require numpy 2.0 to build. I would've had to adjust all those other CI to point them to the wheels bucket.

Instead I modified the wheel build to use `no-build-isolation`.

rgommers commented on .github/workflows/wheels.yml:
```diff
@@ -130,16 +130,25 @@ jobs:
+          CIBW_BEFORE_BUILD_LINUX: "pip install numpy>=2.0.0.dev0 meson-python cython pythran pybind11 ninja; bash {project}/tools/wheels/cibw_before_build_linux.sh {project}"
+          CIBW_BEFORE_BUILD_WINDOWS: "pip install numpy>=2.0.0.dev0 meson-python cython pythran pybind11 ninja && bash {project}/tools/wheels/cibw_before_build_win.sh {project}"
+          CIBW_BEFORE_BUILD_MACOS: "pip install numpy>=2.0.0.dev0 meson-python cython pythran pybind11 ninja; bash {project}/tools/wheels/cibw_before_build_macos.sh {project}"
+          CIBW_BEFORE_TEST: pip install setuptools setuptools_scm[toml] wheel
```
It looks like this is for installing `asv`. I filed an issue on the `asv` repo about missing wheels. However, we should not need `asv` to begin with in our wheel build jobs. It's not immediately clear to me why it's being pulled in.

andyfaff replied:
Asv is not needed in the build step, but during the test step. All wheels are tested during the cibuildwheel process

rgommers replied:
Sure, but `asv` isn't a test dependency - it's for benchmarking, and we don't run benchmarks in the wheel build job.

andyfaff replied:
From looking at the git history it looks like I'm responsible (<chuckling>) for it being added to the [`test-requires` section](https://github.com/scipy/scipy/blob/main/pyproject.toml#L127). However, to be fair I was probably following it's presence in the entries in the [`project.optional-dependencies.test` section](https://github.com/scipy/scipy/blob/main/pyproject.toml#L80).

We can probably remove asv from the former, can we remove it from the latter?

rgommers replied:
Yes. We can put it in a separate `benchmark` section instead I'd say, that'd be more accurate.

rgommers commented:
> Instead I modified the wheel build to use `no-build-isolation`.

Thanks, I understand the problem now. This seems like a good solution.

tylerjereddy commented on .github/workflows/wheels.yml:
```diff
@@ -130,16 +130,25 @@ jobs:
             PIP_PRE=1
-            # PIP_EXTRA_INDEX_URL=https://pypi.anaconda.org/scientific-python-nightly-wheels/simple
+            PIP_EXTRA_INDEX_URL=https://pypi.anaconda.org/scientific-python-nightly-wheels/simple
+            PIP_NO_BUILD_ISOLATION=false
```
Obviously not a SciPy problem, but the naming of this variable/how it is activated is so bad I had to double take--looks like I'm not the only one: https://github.com/pypa/pip/issues/5735

You activate the true meaning of the variable by setting it to a false or null value, which is also compounded by parsing double negative logic when reading it the first time.

andyfaff replied:
Yes, it's pretty bad. 
Note for the future - cibuildwheel is currently unable to send flags/arguments to the build front end. Adding this functionality is being added [here](https://github.com/pypa/cibuildwheel/pull/1588)

rgommers replied:
It'd be useful to put a comment about this _enabling_ build isolation.

tylerjereddy commented:
I think I understand pretty much all the changes here, and sounds like Ralf and I are both "ok" with the approach, and the CI is passing.

I suppose some minor simplifications may be possible to apply re: `asv` still, though I don't really mind if we delay handling that (is it just the deletion of `CIBW_BEFORE_TEST` left in that regard?) until later since there's already a `TODO`-style comment for it that will likely have to change before the next release cycle in late 2023/early 2024.

tylerjereddy commented:
Also, is it 100 % clear what version of NumPy we'd build our PyPI wheels against if `2.0.0` is available at release time in say January 2024? Or is there still too much in motion there?

andyfaff commented:
> is it just the deletion of CIBW_BEFORE_TEST left in that regard?

I removed the CIBW_BEFORE test line. 

> I suppose some minor simplifications may be possible to apply re: asv still,

I think removing `asv` from the `project.optional-dependencies.test` section can be left for a future PR.

> Also, is it 100 % clear what version of NumPy we'd build our PyPI wheels against if 2.0.0 is available at release time in say January 2024? Or is there still too much in motion there?

Ralf pointed out the notes in pyproject.toml about this. My understanding is that if it's available we should build against the 2.0.0 release/RC1, which will be back compatible to 1.25.x. The notes indicate that back to 1.22.x should be possible.

rgommers commented on .github/workflows/wheels.yml:
```diff
@@ -130,16 +130,24 @@ jobs:
+          #  numpy2.0 wheels available on PyPI.
+          # Also remove PIP_NO_BUILD_ISOLATION flag.
+          # For some reason asv needs extra packages for installation on cp312,
+          # but not on lower python versions
```
The two lines above can be removed now, right?

rgommers reviewed (approved):
LGTM, thanks @andyfaff. Leaving the `asv` removal from `optional-dependencies` is fine to not do here. This should be good to merge as it is - although please consider updating the code comments before merging.

andyfaff commented:
@seberg, nightly wheels built against numpy2.0.0.dev0 will start appearing in the nightly bucket in a next day or so.
